File: client/lib/features/tracker/data/models/shared_played_track.dart
class SharedPlayedTrackModel {
  final int id;
  final int internalDeviceId;

  final String deviceId;

  final int trackId;

  final DateTime startAt;
  final DateTime finishAt;

  final Duration beginAt;
  final Duration endedAt;

  final bool shuffle;

  final bool trackEnded;

  final DateTime sharedAt;

  final Duration trackDuration;

  const SharedPlayedTrackModel({
    required this.id,
    required this.internalDeviceId,
    required this.deviceId,
    required this.trackId,
    required this.startAt,
    required this.finishAt,
    required this.beginAt,
    required this.endedAt,
    required this.shuffle,
    required this.trackEnded,
    required this.trackDuration,
    required this.sharedAt,
  });

  factory SharedPlayedTrackModel.fromJson(Map<String, dynamic> json) {
    return SharedPlayedTrackModel(
      id: json['id'] as int,
      internalDeviceId: json['internal_device_id'] as int,
      deviceId: json['device_id'] as String,
      trackId: json['track_id'] as int,
      startAt: DateTime.parse(json['start_at'] as String),
      finishAt: DateTime.parse(json['finish_at'] as String),
      beginAt: Duration(milliseconds: json['begin_at'] as int),
      endedAt: Duration(milliseconds: json['ended_at'] as int),
      shuffle: json['shuffle'] as bool,
      trackEnded: json['track_ended'] as bool,
      trackDuration: Duration(milliseconds: json['track_duration'] as int),
      sharedAt: DateTime.parse(json['shared_at'] as String),
    );
  }
}

File: server/internal/layers/data/models/shared_played_track.go
package data_models

import (
	"time"

	"github.com/gungun974/Melodink/server/internal/layers/domain/entities"
)

type SharedPlayedTrackModels []SharedPlayedTrackModel

func (s SharedPlayedTrackModels) ToSharedPlayedTracks() []entities.SharedPlayedTrack {
	e := make([]entities.SharedPlayedTrack, 0, len(s))

	for _, m := range s {
		e = append(e, m.ToSharedPlayedTrack())
	}

	return e
}

type SharedPlayedTrackModel struct {
	Id               int `db:"id"`
	InternalDeviceId int `db:"internal_device_id"`

	UserId   int    `db:"user_id"`
	DeviceId string `db:"device_id"`

	TrackId int `db:"track_id"`

	StartAt  time.Time `db:"start_at"`
	FinishAt time.Time `db:"finish_at"`

	BeginAt int `db:"begin_at"`
	EndedAt int `db:"ended_at"`

	Shuffle       bool `db:"shuffle"`
	TrackEnded    bool `db:"track_ended"`
	TrackDuration int  `db:"track_duration"`

	CreatedAt time.Time  `db:"created_at"`
	UpdatedAt *time.Time `db:"updated_at"`
}

func (m *SharedPlayedTrackModel) ToSharedPlayedTrack() entities.SharedPlayedTrack {
	sharedAt := m.CreatedAt

	if m.UpdatedAt != nil {
		sharedAt = *m.UpdatedAt
	}

	return entities.SharedPlayedTrack{
		Id:               m.Id,
		InternalDeviceId: m.InternalDeviceId,

		UserId:   m.UserId,
		DeviceId: m.DeviceId,

		TrackId: m.TrackId,

		StartAt:  m.StartAt,
		FinishAt: m.FinishAt,

		BeginAt: m.BeginAt,
		EndedAt: m.EndedAt,

		Shuffle:       m.Shuffle,
		TrackEnded:    m.TrackEnded,
		TrackDuration: m.TrackDuration,

		SharedAt: sharedAt,
	}
}
